### PR TITLE
in_forward: Add source_hostname_key parameter. fix #804

### LINKED ISSUE #804: Add option to enable source host key for in_forward plugin
Like other plugins, it would a really helpful to be able to use "source_host_key" for in_forward as is possible with in_tcp.

### PR DESCRIPTION
**Pull request by repeatedly**

### CODE REVIEW

File: lib/fluent/plugin/in_forward.rb [hunk 1]
```diff
@@ -47,6 +47,8 @@ def initialize
     config_param :chunk_size_limit, :size, default: nil
     desc 'Skip an event if incoming event is invalid.'
     config_param :skip_invalid_event, :bool, default: false
+    desc "The field name of the client's hostname."
```
Comment (tagomoris) [inline lib/fluent/plugin/in_forward.rb]: IMO it's better to have a not for performance degradation.

Comment (repeatedly) [inline lib/fluent/plugin/in_forward.rb]: It seems document issue like this: https://github.com/fluent/fluentd-docs/pull/215/files

**Comment by tagomoris** [comment]
This feature will be supported by porting into v0.14 API and using `inject` plugin helper.

**Comment by repeatedly** [comment]
@tagomoris How to insert client's hostname by inject plugin helper?

**Comment by repeatedly** [comment]
Rebased.

**Comment by tagomoris** [comment]
Oh, I misunderstood the problem.

**Comment by tagomoris** [comment]
IMO, parameter name should be `source_hostname_key` instead of `source_host_key` to show this feature injects hostname, not host address or others explicitly.

**Comment by repeatedly** [comment]
It is derived from `in_tcp` so you also think in_tcp's parameter should be renamed?

**Comment by tagomoris** [comment]
Adding newer name to `in_tcp` sounds good idea.
`in_forward` plugin is used much widely than `in_tcp`, so I think we should NOT follow the bad way of `in_tcp`.

**Comment by repeatedly** [comment]
Renamed.

File: test/plugin/test_in_forward.rb [hunk 2]
```diff
@@ -583,5 +583,53 @@ def send_data(data, try_to_receive_response=false, response_timeout=1)
...
+    }
+  end
+
+  test 'forward protocol with source_hostname_key' do
+    execute_test { |events|
+      entries = []
+      events.each {|tag,time,record|
+        entries << [time, record]
+      }
+      send_data ['tag1', entries].to_msgpack
```
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: I cannot understand why this code works well... `time` should be an ext typed value, but there are no info about it.

File: test/plugin/test_in_forward.rb [hunk 3]
```diff
@@ -583,5 +583,53 @@ def send_data(data, try_to_receive_response=false, response_timeout=1)
...
+      ["tag1", time, {"a"=>1}],
+      ["tag1", time, {"a"=>2}]
+    ]
+    d.expected_emits_length = events.length
+
+    d.run do
+      block.call(events)
+    end
+
+    d.emits.each { |tag, _time, record|
```
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: Does `d.emits` really have 2 events? This test code doesn't wait/confirm to receive data via TCP socket.
I doubt that `d.emits` is empty here.

File: test/plugin/test_in_forward.rb (hunk 2 above)
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: Correct way to do it is using `Fluent::MessagePackFactory`.

File: test/plugin/test_in_forward.rb (hunk 3 above)
Comment (repeatedly) [inline test/plugin/test_in_forward.rb]: Test calls `d.expected_emits_length = events.length` before so `d.run` is finished after received 2 events.

File: test/plugin/test_in_forward.rb (hunk 2 above)
Comment (repeatedly) [inline test/plugin/test_in_forward.rb]: EventTime is serialized to int via `to_msgpack` and this is why this test works.

Comment (repeatedly) [inline test/plugin/test_in_forward.rb]: Yeah, using Fluent::MessagePackFactory is better. "packed forward protocol with source_hostname_key" case does it.

Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: It looks incorrect behavior to simulate `out_forward`... but not so serious for this test.

**Comment by repeatedly** [comment]
Okay. After test passed, I will merge this patch.

**Pull request merged**